### PR TITLE
[vernac] expose API to elaborate declarations

### LINKED ISSUE #14737: Package configuration of inductives in vernac/record.ml
As noticed in #14699, the configuration parameters for many functions about records are passed around seperately. We ought to package these together in some sort of record for readability and maintainability.

### PR DESCRIPTION
**Pull request by gares**
In the context of Elpi I want to let coq elaborate an inductive declaration (up to the constr level) without necessarily adding it to the environment, the same for records and context.
This PR exports the first half of do_mutual_inductive (the part which is side effect free) and do_record and do_context.

Fixes #14737

TODO:
- [x] overlay for coq-elpi using the new API

### CODE REVIEW

File: vernac/comInductive.mli
Comment (ejgallego) [inline vernac/comInductive.mli:53]: Please put this into a record so fields are qualified. `MInd_declaration.t` maybe?

**Comment by ejgallego** [comment]
Modulo the two nits this looks good to me.

**Review by ejgallego — APPROVED** [review]
LGTM, @gares I assume you want to wait for the overlay, so I will wait to merge.

**Comment by gares** [comment]
I'm not sure the code the last commit exposes is in the right place. Happy to move it elsewhere.

File: vernac/vernacentries.mli
Comment (ejgallego) [inline vernac/vernacentries.mli:51]: Don't we have an alias for this long type?

File: vernac/vernacentries.ml
Comment (ejgallego) [inline vernac/vernacentries.ml:984]: I'd move the dumpglob code to a separate function too.

File: vernac/vernacentries.mli
Comment (gares) [reply vernac/vernacentries.mli]: done

**Comment by gares** [comment]
So, after more work on the overlay, the exposed API is:
- pre-processing of the data carried by VernacInductive, which gives two disctinct AST: inductive or record
- elaboration of inductive
- elaboration of record

The latter one has a very nasty type, suggestions are welcome

**Comment by ejgallego** [comment]
Thanks @gares will have a deeper look when the overlay is ready, but indeed I agree with your comments, we can do a lot of improvements in this area. For sure this PR will improve things already.

**Comment by gares** [comment]
I'm almost done with the overlay, sorry for taking so long, but it was not trivial. So far the API is sufficient, but I'm not done.

**Comment by gares** [comment]
OK, so the overlay is done (but I need more cleanup/doc before merging it, so I did not put the PR up there) and the API I do export are sufficient to let an external tool interpret (from expr to constr) all declarations:
- it was already the case for definition/axiom
- it is now the case also for inductive/record and context

The PR is ready for another round of review.

**Comment by gares** [comment]
ok, then ping @coq/vernac-maintainers, this is not a very tough one.

File: vernac/record.mli
Comment (ejgallego) [reply vernac/record.mli:48]: Yes, I'm not sure the current stuff in `record.ml` is suited for export, see also https://github.com/coq/coq/pull/15877 [which I still don't quite follow sadly]

File: vernac/comAssumption.mli
Comment (gares) [inline vernac/comAssumption.mli:67]:
```suggestion
  -> Environ.env
```

File: vernac/record.mli
Comment (gares) [reply vernac/record.mli:39]: done

File: vernac/record.ml
Comment (Alizter) [inline vernac/record.ml:580]:
```suggestion
    | Vernacexpr.AssumExpr({CAst.v=Name id},_,_) -> id::acc
```

**Review by Alizter — COMMENTED** [review]
I guess this also closes https://github.com/coq/coq/issues/14737

File: vernac/comAssumption.mli
Comment (ejgallego) [inline vernac/comAssumption.mli:65]: I have a bit of confusion with terminology here, so usally we call interp to the phase constexpr to glob_constr, however here we are also calling pretyping. The sigma also play a role, but seems to be needed by accident, so maybe we can put a comment "cleanup sigma uses in do_context"?

Comment (ejgallego) [reply vernac/comAssumption.mli:65]: Regarding the terminology I'm fine with whatever we choose, but IMHO we should use something different from interp for what this is which is interp+pretype.

File: vernac/comAssumption.mli
Comment (ejgallego) [inline vernac/comAssumption.mli:69]: That return type could use a record too, note my comment below on the return evar appearing here to be an accident.

File: vernac/record.mli
Comment (ejgallego) [inline vernac/record.mli:55]: Would be good to put a comment here mentioning a todo to further refactor this with the code for inductives, which I believe it can be done.

**Review by ejgallego — APPROVED** [review]
Looks OK, tho I'd be nice to add some clarifications in form of comments, as I noted in my review.

File: vernac/record.mli
Comment (Alizter) [reply vernac/record.mli:55]: I suppose that the goal of this PR is not really to refactor the vernac for records to much as it was to expose API that Enrico wished to use. When investigating how to go about https://github.com/coq/coq/issues/14737 I ran into a few problems, mainly that it is not entirely clear what the data of a record actually is. Many functions require only partial bits of the record data, so it is not so obvious how to organize this. I don't think that will be solved in this PR, so I second putting a comment for the future here.

File: vernac/comAssumption.mli
Comment (SkySkimmer) [reply vernac/comAssumption.mli:65]: >usally we call interp to the phase constexpr to glob_constr

No, that's intern. see eg Constrintern.intern_constr and Constrintern.interp_constr

**Comment by gares** [comment]
I believe this is OK now. My overlay is complete, so I tested the feasibility of using the exposed APIs

**Comment by gares** [comment]
fiat failure seems unrelated. I can't say about analysis.

**Comment by gares** [comment]
CI was green but for fiat which is fixed in master, so I did rebase to get a fully green CI.
After that, please merge.

**Comment by gares** [comment]
@SkySkimmer would you mind taking care of this? I'd like to first merge this elpi overlay, and then prepare the one for #15946

**Comment by SkySkimmer** [comment]
@ejgallego is still assigned though

**Comment by Alizter** [comment]
@SkySkimmer I think it is OK to take over in this case. The assignment was a month ago, so I don't think we can reasonably expect it to be looked at.

**Comment by SkySkimmer** [comment]
Last activity was just before holidays though

**Comment by gares** [comment]
I think @ejgallego was OK with the PR both in principle and with the current code, I did only add a few comments after his review.

**Comment by ejgallego** [comment]
> The assignment was a month ago, so I don't think we can reasonably expect it to be looked at.

I am not sure why, my last activity was indeed 11 days ago, then I went on holidays [and I did a review but the PR was not ready, so really I could only have reacted on the weekend]

That being said, the PR looks good to me, so go ahead and take over. A think to keep in mind tho is the conflict with #15877 , that one I did have a look but I didn't manage to understand the code. I'd suggest pinging @proux01 in case he'd like to comment anything about this one.

File: vernac/comAssumption.mli
Comment (ejgallego) [reply vernac/comAssumption.mli:65]: Yes, I get confused by the directory name being `interp`, interp is the compostion of intern + pretyping.

Maybe we could indeed think about the terminology.

**Comment by gares** [comment]
If you are back from VAC you can also merge it yourself. I'd like to merge overlay, since I then have to make another one which will generate conflicts, so a quick merge would be appreciated

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**
